catalog/spatial/registry/registry-publication-action-provider/src/main/java/org/codice/ddf/registry/publication/action/provider/RegistryPublicationActionProvider.java

```diff
@@ -0,0 +1,262 @@
+            }
+        }
+    }
+
+    private Action getAction(String regId, String destinationId, boolean publish) {
+
+        URL url;
+        String title = publish ? PUBLISH_TITLE + destinationId : UNPUBLISH_TITLE + destinationId;
+        String description = publish ?
+                PUBLISH_DESCRIPTION + destinationId :
+                UNPUBLISH_DESCRIPTION + destinationId;
+        String operation = publish ? PUBLISH_OPERATION : UNPUBLISH_OPERATION;
+        String httpOp = publish ? HTTP_POST : HTTP_DELETE;
+        try {
+
+            String path = String.format("%s/%s/%s/%s",
+                    REGISTRY_PATH,
+                    regId,
+                    PUBLICATION_PATH,
+                    URLEncoder.encode(destinationId, "UTF-8"));
+            URI uri = new URI(SystemBaseUrl.constructUrl(path, true));
+            url = uri.toURL();
+
+        } catch (MalformedURLException | URISyntaxException | UnsupportedEncodingException e) {
+            LOGGER.error("Malformed URL exception", e);
+            return null;
+        }
+        String id = String.format("%s.%s.%s", getId(), operation, httpOp);
+        return new ActionImpl(id, title, description, url);
+    }
+
+    private <T> String getRegistryId(T subject) {
+        if (subject instanceof Metacard) {
+            Metacard mcard = (Metacard) subject;
+            if (!mcard.getTags()
+                    .contains(RegistryConstants.REGISTRY_TAG)) {
+                return null;
+            }
+            return ((Metacard) subject).getAttribute(RegistryObjectMetacardType.REGISTRY_ID)
+                    .getValue()
+                    .toString();
+        } else if (subject instanceof Source) {
+            try {
+                Configuration[] configurations = configAdmin.listConfigurations(String.format(
+                        "(id=%s)",
+                        ((Source) subject).getId()));
+                return (String) Arrays.stream(configurations)
+                        .map(Configuration::getProperties)
+                        .map(p -> p.get(RegistryObjectMetacardType.REGISTRY_ID))
+                        .filter(Objects::nonNull)
+                        .findFirst()
+                        .orElse(null);
+
+            } catch (IOException | InvalidSyntaxException e) {
+                LOGGER.warn("Couldn't get configuration for {}", ((Source) subject).getId(), e);
+            }
+        } else if (subject instanceof Configuration) {
+            return (String) ((Configuration) subject).getProperties()
+                    .get(RegistryObjectMetacardType.REGISTRY_ID);
+        }
+        return null;
+    }
+
+    public void setProviderId(String id) {
+        this.providerId = id;
+    }
+
+    public void setConfigAdmin(ConfigurationAdmin admin) {
+        this.configAdmin = admin;
+    }
+
+    public void setRegistryStores(List<RegistryStore> registryStores) {
+        this.registryStores = registryStores;
```

Comment:
This list could be changed as we're iterating over it in the `getActions()` method. May want to use a `CopyOnWriteArrayList`, clear it and add the new elements, or synchronize this method and the `getActions()` for loop.